src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java

```diff
@@ -0,0 +1,202 @@
+package at.ac.tuwien.kr.alpha.common.depgraph;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.grounder.NonGroundRule;
+
+// TODO ensure immutability
+public class DependencyGraph {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(DependencyGraph.class);
+
+	private static final String CONSTRAINT_PREDICATE_FORMAT = "[constr_%d]";
+
+	/**
+	 * Maps nodes to outgoing edges of that node NOTE: Doing the value as List<Edge> rather than Map<Integer,Boolean> as one node may have positive and negative
+	 * edges to another.
+	 */
+	private Map<Node, List<Edge>> nodes = new HashMap<>();
+
+	private Map<Predicate, Node> nodesByPredicate = new HashMap<>();
+
+	/**
+	 * The transposed graph structure, i.e. the same set of nodes, but all edges reversed - needed for analysis of strongly connected components.
+	 */
+	private Map<Node, List<Edge>> transposedNodes;
+
+	private Map<Integer, List<Node>> stronglyConnectedComponents;
+
+	private int constraintNumber;
+
+	public DependencyGraph() {
+		// doing this since checkstyle doesn't let me initialize it directly to 0
```

Comment:
Checkstyle doesn't let you, because `0` is the default initialization value. If you put nothing it will be `0` anyways. This default constructor thus can be removed without changing anything.